esp-hal/src/i2c.rs

```diff
@@ -885,160 +969,186 @@ mod asynch {
+            for (idx, chunk) in read_buffer.chunks_mut(I2C_CHUNK_SIZE).enumerate() {
+                // Clear all I2C interrupts
+                self.peripheral.clear_all_interrupts();
+
+                let cmd_iterator = &mut self.peripheral.register_block().comd_iter();
+
+                self.read_operation(
+                    address,
+                    chunk,
+                    idx == 0,
+                    idx == chunk_count - 1,
+                    idx < chunk_count - 1,
+                    cmd_iterator,
+                )
+                .await?;
+            }
+
             Ok(())
         }
-    }
 
-    impl<'d, T> embedded_hal_async::i2c::I2c for I2C<'d, T, crate::Async>
-    where
-        T: Instance,
-    {
+        /// Execute the provided operations on the I2C bus as a single
+        /// transaction.
+        ///
+        /// Transaction contract:
+        /// - Before executing the first operation an ST is sent automatically.
+        ///   This is followed by SAD+R/W as appropriate.
+        /// - Data from adjacent operations of the same type are sent after each
+        ///   other without an SP or SR.
+        /// - Between adjacent operations of a different type an SR and SAD+R/W
+        ///   is sent.
+        /// - After executing the last operation an SP is sent automatically.
+        /// - If the last operation is a `Read` the master does not send an
+        ///   acknowledge for the last byte.
+        ///
+        /// - `ST` = start condition
+        /// - `SAD+R/W` = slave address followed by bit 1 to indicate reading or
+        ///   0 to indicate writing
+        /// - `SR` = repeated start condition
+        /// - `SP` = stop condition
         async fn transaction(
             &mut self,
             address: u8,
             operations: &mut [Operation<'_>],
-        ) -> Result<(), Self::Error> {
-            let mut last_op = Op::None;
+        ) -> Result<(), Error> {
+            let mut last_op: Option<OpKind> = None;
             // filter out 0 length read operations
             let mut op_iter = operations
                 .iter_mut()
-                .filter(|op| match op {
-                    Operation::Write(_) => true,
-                    Operation::Read(buffer) => !buffer.is_empty(),
-                })
+                .filter(|op| if op.is_write() { true } else { !op.is_empty() })
                 .peekable();
+
             while let Some(op) = op_iter.next() {
-                let next_op: Op = op_iter.peek().into();
+                let next_op = op_iter.peek().map(|v| v.kind());
+                let kind = op.kind();
                 // Clear all I2C interrupts
                 self.peripheral.clear_all_interrupts();
 
                 let cmd_iterator = &mut self.peripheral.register_block().comd_iter();
-                match op {
-                    Operation::Write(bytes) => {
+                match kind {
```

Comment:
Could we match on `op` instead, and avoid the `unwrap`s?

Reply:
well the function's signature was wrong and it should be `operations: &mut [impl I2cOperation]` and then I can't - I'm not too happy with the whole I2cOperation thing but it was the only way to overcome lifetime problems

The only other way I see would be to have the inherent function take `operations: &mut [Operation]` and duplicate the code in the eh impl taking `operations: &mut [embedded_hal::i2c::Operation<'_>]`

Reply:
Can we turn this into `operations: &mut [impl Into<OurI2cOperationType>]`, convert and _then_ match?

Reply:
can't get it to work - in the end there are lifetime issues

Reply:
Shame.